examples/tree_source/tree_source/app.py

```diff
@@ -1,119 +1,157 @@
-from random import choice
+from pathlib import Path
 
 import toga
 from toga.style import Pack
-from toga.constants import ROW, COLUMN
+from toga.constants import COLUMN
 from toga.sources import Source
+from datetime import datetime
 
-bee_movies = [
-    {'year': 2008, 'title': 'The Secret Life of Bees', 'rating': '7.3', 'genre': 'Drama'},
-    {'year': 2007, 'title': 'Bee Movie', 'rating': '6.1', 'genre': 'Animation, Adventure, Comedy'},
-    {'year': 1998, 'title': 'Bees', 'rating': '6.3', 'genre': 'Horror'},
-    {'year': 2007, 'title': 'The Girl Who Swallowed Bees', 'rating': '7.5', 'genre': 'Short'},
-    {'year': 1974, 'title': 'Birds Do It, Bees Do It', 'rating': '7.3', 'genre': 'Documentary'},
-    {'year': 1998, 'title': 'Bees: A Life for the Queen', 'rating': '8.0', 'genre': 'TV Movie'},
-    {'year': 1994, 'title': 'Bees in Paradise', 'rating': '5.4', 'genre': 'Comedy, Musical'},
-    {'year': 1947, 'title': 'Keeper of the Bees', 'rating': '6.3', 'genre': 'Drama'}
-]
-
-class Movie:
-    # A class to wrap individual
-    def __init__(self, year, title, rating, genre):
-        self.year = year
-        self.title = title
-        self.rating = rating
-        self.genre = genre
-
-
-class Decade:
-    # A class to wrap
-    def __init__(self, decade):
-        self.decade = decade
-        self._data = []
-
-    # Display values for the decade in the tree.
-    @property
-    def year(self):
-        return "{}0's".format(self.decade)
+# This is a slightly less toy example of a tree view to display
+# folders in your home directory. To avoid loading everything
+# in advance, the subfolders are loaded dynamically, on access.
+# This typically happens when the tree view asks for the number
+# of children to decide if it should display an expandable row.
+#
+# In practice, one can use a similar concept of on-access
+# loading to fetch data from an API or online resource but
+# perform the loading asynchronously.
+
+
+class LoadingFailedNode:
+    """A node to represent failed loading of children"""
+
+    def __init__(self, parent):
+        self.parent = parent
+        self.children = []
+        self.name = 'loading failed'
+        self.date_modified = ''
 
     # Methods required for the data source interface
     def __len__(self):
-        return len(self._data)
+        return 0
 
     def __getitem__(self, index):
-        return self._data[index]
+        pass
```

Comment:
This needs to `raise StopIteration`. If you try print `[n for n in LoadingFailedNode()]`, you'll cause your Python interpreter to crash, as the iteration protocol doesn't use `__len__` to determine how many items are in the list. Not sure why this doesn't cause problems on Cocoa, but it causes GTK to crash (possibly because GTK doesn't do fully deferred loading)

Reply:
Good point. I had overlooked this, probably because it did not cause a crash in Cocoa.